WS master should send plugin IDs instead of plugins metas to unified broker

amisevsk opened the pull request:
### What does this PR do?
Makes ws master send plugin FQNs to broker instead of downloaded meta.yamls.

Draft PR until https://github.com/eclipse/che-plugin-broker/pull/40 is merged.

### What issues does this PR fix or reference?
https://github.com/redhat-developer/rh-che/issues/1307; related to https://github.com/eclipse/che/issues/12908

garagatyi commented on wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  @Test
+  public void shouldParseAllPluginsAndEditor() throws Exception {
```
I know that you took these cases from another test class. But could you rework this test to test with data provider? It would looks way simpler and could include all the non exception cases like what you have in `shouldReturnEmptyListWhenNoPluginsOrEditors`.
Apart from making code more concise and simplify its support you could split cases where different bugs would not fail the same test case as it is now with this test. 
E.g. 
- case with registry
- case with registry that has path /plugins
- case with registry that has path not equal /plugins
- case with registry that has multi-level path
- case without registry
- case with null editor
- case with null plugins
- case with multiple plugins
- etc

To simplify matching of results you can use `assertEqualsnoOrder(actual.toArray(), expected.toArray())` (please excuse not precise names of methods if any)

garagatyi commented:
Please take into account that after merging this PR we will need to adapt RhChe https://github.com/redhat-developer/rh-che/pull/1192/files#diff-d1c7cee2e326bd168cb76031bef4ef0cR197

amisevsk commented:
Marked as ready for review since it works with locally built plugin-broker, but should not be released until https://github.com/eclipse/che-plugin-broker/pull/46 is merged and a new image is available.

amisevsk commented:
For testing, images `amisevsk/che-unified-plugin-broker:v0.15.2` and `amisevsk/che-init-plugin-broker:v0.15.2` can be used.

sleshchenko commented on wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParserTest.java:
```diff
@@ -0,0 +1,207 @@
+
+  @Test(
+      expectedExceptions = InfrastructureException.class,
+      expectedExceptionsMessageRegExp = "Invalid.*duplicated")
```
The full expected message is `String.format("Invalid Che tooling plugins configuration: plugin %s:%s is duplicated", ...)` and I think it makes sense to check full message to make sure that message is formatted correctly (message contains `ID:VERSION` but not other values). I started to ask it after I saw that Che Server respond with errors where messages are wrong formatted, and it's pretty easy to make mistake here especially during refactoring.
WDYT?
It makes less sense to check the full message that is not formatted but still makes sense. It's like double fields for putting a password (password, confirm password). Also, if the message becomes formatted - then the author should change tests accordingly, otherwise it would be changed at all.

amisevsk replied:
Yeah it makes sense, I usually default to keeping the match as general as possible (only to ensure exception is being thrown for the right reason) but validating that plugin id and version are included makes a lot of sense.

sleshchenko commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java:
```diff
@@ -0,0 +1,103 @@
+
+    List<PluginFQN> metaFQNs = new ArrayList<>();
+    if (!isNullOrEmpty(pluginsAttribute)) {
+      String[] plugins = pluginsAttribute.split(" *, *");
```
` *, *` makes sure that empty spaces will be removed between plugins but it does not remove empty spaces before the first plugin and after the last plugin. Maybe it would be safer to split by `,` and then trim split values.
```java
  @Test(
      expectedExceptions = InfrastructureException.class,
      expectedExceptionsMessageRegExp = "Invalid.*duplicated")
  public void shouldDetectDuplicatedPluginsIfTheyArePrefixedSuffixedWithEmptySpaces() throws Exception {
    Map<String, String> attributes =
        createAttributes(
            "",
            " " + formatPlugin("http://testregistry1:8080", "testplugin", "1.0"),
            formatPlugin("http://testregistry2:8080", "testplugin", "1.0") + " ");

    parser.parsePlugins(attributes);
  }
```

amisevsk replied:
Good catch -- done. And thanks for the test case, I've added it to the suite.

sleshchenko commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java:
```diff
@@ -53,4 +57,28 @@ public String getVersion() {
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```
Please add `toString` as well. It may be useful in tests

amisevsk replied:
Done.

sleshchenko commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/model/PluginFQN.java:
```diff
@@ -53,4 +57,28 @@ public String getVersion() {
   }
+
+  @Override
+  public int hashCode() {
```
`hashCode` might look simpler like
```java
  @Override
  public int hashCode() {
    return Objects.hash(getRegistry(), getId(), getVersion());
  }
```
But I'm OK with current way as well.

amisevsk replied:
The pitfalls of auto-generated code. Fixed.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java:
```diff
@@ -65,10 +67,10 @@
 public abstract class BrokerEnvironmentFactory<E extends KubernetesEnvironment> {
 
-  private static final String CONFIG_MAP_NAME_SUFFIX = "broker-config-map";
+  @VisibleForTesting protected static final String CONFIG_MAP_NAME_SUFFIX = "broker-config-map";
```
Good point, moved to package private.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java:
```diff
@@ -106,7 +111,7 @@ public BrokerEnvironmentFactory(
    */
   public E create(
-      Collection<PluginMeta> pluginsMeta, RuntimeIdentity runtimeID, BrokersResult brokersResult)
+      Collection<PluginFQN> pluginFQNs, RuntimeIdentity runtimeID, BrokersResult brokersResult)
```
Updated java doc and removed brokersResult -- good catch that must be leftover from the merging of brokers.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/brokerphases/BrokerEnvironmentFactory.java:
```diff
@@ -87,13 +90,15 @@ public BrokerEnvironmentFactory(
       MachineTokenEnvVarProvider machineTokenEnvVarProvider,
       @Named("che.workspace.plugin_broker.unified.image") String unifiedBrokerImage,
-      @Named("che.workspace.plugin_broker.init.image") String initBrokerImage) {
+      @Named("che.workspace.plugin_broker.init.image") String initBrokerImage,
```
Tha's what I get for writing code to fit what's already there :)

Removed annotations.

amisevsk commented on wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginFQNParser.java:
```diff
@@ -0,0 +1,103 @@
+@Beta
+public class PluginFQNParser {
+
+  public Collection<PluginFQN> parsePlugins(Map<String, String> attributes)
```
Added docs

amisevsk commented:
ci-test

amisevsk commented:
Waiting on a release of Che plugin broker to mirror the necessary changes.

**Do not merge**

garagatyi commented:
@amisevsk I see that there is a release with necessary changes but it is done manually instead of by CI and container image is not pushed to Docker hub. Do you plan to push them? Do you need help with that?

amisevsk commented:
@garagatyi Thanks for the help, I couldn't figure out the CI backing plugin-broker releases. 

Bumped broker version in che.properties, squashed all commits, and rebased on current master. Tested locally and everything works with FQN changes.
